### PR TITLE
Add 24 bit dictionary support to Parquet writer

### LINKED ISSUE #10948: [FEA] Support dictionary compression with up to 32 bits for Parquet writer
Currently, the parquet writer has a 16 bit maximum dictionary size.

https://github.com/rapidsai/cudf/blob/da7474471997f31cf16d552f3ae8f74ea10390b0/cpp/src/io/parquet/parquet_gpu.hpp#L48-L50

This means that within a column chunk, if there are <= 65535 unique entries, we can use dictionary based compression to potentially save space. However, if there are > 65535 unique entries we can't use dictionary compression, potentially leading to larger file sizes.

The [Apache Parquet format specification indicates support for a 32 bit maximum width](https://github.com/apache/parquet-format/blob/master/Encodings.md#dictionary-encoding-plain_dictionary--2-and-rle_dictionary--8), and this is available in the [official Java implementation of Parquet](https://github.com/apache/parquet-mr).

Some users have shared that using libcudf instead of the traditional Java Parquet implementation has resulted in file sizes >5x larger in some scenarios (due to not being able to use dictionary encoding), which can pose a storage challenge at scale.

### PR DESCRIPTION
**Pull request by devavret**
Closes #10948 

Adds support for dictionary encoding with 24 bit indices.

### CODE REVIEW

File: cpp/src/io/parquet/page_enc.cu [hunk 1]
```diff
@@ -505,11 +506,18 @@ inline __device__ void PackLiterals(
...
+          dst[t * 3 + 1] = v >> 8;
+          dst[t * 3 + 2] = v >> 16;
+        }
+        return;
       }
     } else {
       return;
     }
-  } else {
+  } else if (w <= 16) {
```
Comment (davidwendt) [inline cpp/src/io/parquet/page_enc.cu:520]: How does this else-if get reached over the if clause itself? Are there other values for `w` in between the ones listed in the if-statement `(w == 1 || w == 2 || w == 4 || w == 8 || w == 12 || w == 16 || w == 24)`? Or perhaps I'm reading this wrong.

Comment (devavret) [reply cpp/src/io/parquet/page_enc.cu:520]: That is correct. For a limited number of values, there's a faster method with a shfl or simply byte copies. The rest need atomics.

**Review by davidwendt — APPROVED** [review]
Looks good to me.

**Comment by vuule** [comment]
@devavret please add description to the PR

**Review by vuule — CHANGES_REQUESTED** [review]
Just one small suggestion, LGTM otherwise.

**Comment by PointKernel** [comment]
@devavret You mentioned there was a performance penalty by expanding the dictionary to 24 bits. Is the problem solved?

**Comment by devavret** [comment]
> @devavret You mentioned there was a performance penalty by expanding the dictionary to 24 bits. Is the problem solved?

Yes. The map insert kernel increases in time. But it is entirely compensated by the reduction in IO time caused by a reduced file size.

**Comment by vuule** [comment]
@devavret do you have the benchmark comparison? I think you mentioned there's a big difference in both memory use and in performance.

**Review by vuule — APPROVED** [review]

**Comment by devavret** [comment]
Here are the results. Notice the `encoded_file_size` UserCounter is smaller even for large cardinality.

```
Int64

Existing
------------------------------------------------------------------------------------------------------------------------
Benchmark                                                              Time             CPU   Iterations UserCounters...
------------------------------------------------------------------------------------------------------------------------
ParquetWrite/integral_file_output/4/1000/1/0/0/manual_time           103 ms         82.8 ms            6 bytes_per_second=4.86455G/s encoded_file_size=102.715M peak_memory_usage=690.322M
ParquetWrite/integral_file_output/4/262143/1/0/0/manual_time         446 ms          362 ms            2 bytes_per_second=1.12165G/s encoded_file_size=534.274M peak_memory_usage=542.542M
ParquetWrite/integral_file_output/4/524288/1/0/0/manual_time         422 ms          359 ms            2 bytes_per_second=1.18515G/s encoded_file_size=534.248M peak_memory_usage=542.497M
ParquetWrite/integral_file_output/4/1048576/1/0/0/manual_time        443 ms          360 ms            2 bytes_per_second=1.12965G/s encoded_file_size=534.218M peak_memory_usage=542.453M
ParquetWrite/integral_file_output/4/2097152/1/0/0/manual_time        421 ms          360 ms            2 bytes_per_second=1.18725G/s encoded_file_size=534.21M peak_memory_usage=542.442M
ParquetWrite/integral_file_output/4/4194304/1/0/0/manual_time        489 ms          368 ms            2 bytes_per_second=1047.73M/s encoded_file_size=534.19M peak_memory_usage=542.407M

With 24 bit
------------------------------------------------------------------------------------------------------------------------
Benchmark                                                              Time             CPU   Iterations UserCounters...
------------------------------------------------------------------------------------------------------------------------
ParquetWrite/integral_file_output/4/1000/1/0/0/manual_time           126 ms         91.4 ms            7 bytes_per_second=3.96883G/s encoded_file_size=102.715M peak_memory_usage=1.42143G
ParquetWrite/integral_file_output/4/262143/1/0/0/manual_time         313 ms          284 ms            2 bytes_per_second=1.59719G/s encoded_file_size=322.962M peak_memory_usage=1.42143G
ParquetWrite/integral_file_output/4/524288/1/0/0/manual_time         430 ms          369 ms            2 bytes_per_second=1.16204G/s encoded_file_size=415.222M peak_memory_usage=1.42143G
ParquetWrite/integral_file_output/4/1048576/1/0/0/manual_time        524 ms          443 ms            2 bytes_per_second=976.848M/s encoded_file_size=517.408M peak_memory_usage=1.39121G
ParquetWrite/integral_file_output/4/2097152/1/0/0/manual_time        585 ms          409 ms            2 bytes_per_second=875.166M/s encoded_file_size=534.21M peak_memory_usage=769.133M
ParquetWrite/integral_file_output/4/4194304/1/0/0/manual_time        569 ms          397 ms            1 bytes_per_second=900.239M/s encoded_file_size=534.19M peak_memory_usage=769.133M


String
Existing
-----------------------------------------------------------------------------------------------------------------------
Benchmark                                                             Time             CPU   Iterations UserCounters...
-----------------------------------------------------------------------------------------------------------------------
ParquetWrite/string_file_output/23/1000/1/0/0/manual_time          70.4 ms         59.0 ms           10 bytes_per_second=7.10482G/s encoded_file_size=51.8321M peak_memory_usage=345.687M
ParquetWrite/string_file_output/23/262143/1/0/0/manual_time         580 ms          448 ms            1 bytes_per_second=882.824M/s encoded_file_size=661.238M peak_memory_usage=665.42M
ParquetWrite/string_file_output/23/524288/1/0/0/manual_time         569 ms          459 ms            1 bytes_per_second=899.767M/s encoded_file_size=661.256M peak_memory_usage=665.431M
ParquetWrite/string_file_output/23/1048576/1/0/0/manual_time        536 ms          447 ms            1 bytes_per_second=955.749M/s encoded_file_size=661.205M peak_memory_usage=665.369M
ParquetWrite/string_file_output/23/2097152/1/0/0/manual_time        538 ms          449 ms            1 bytes_per_second=952.19M/s encoded_file_size=661.218M peak_memory_usage=665.379M
ParquetWrite/string_file_output/23/4194304/1/0/0/manual_time        537 ms          448 ms            1 bytes_per_second=953.421M/s encoded_file_size=661.274M peak_memory_usage=665.431M

with 24 bit
-----------------------------------------------------------------------------------------------------------------------
Benchmark                                                             Time             CPU   Iterations UserCounters...
-----------------------------------------------------------------------------------------------------------------------
ParquetWrite/string_file_output/23/1000/1/0/0/manual_time          64.1 ms         54.7 ms           10 bytes_per_second=7.79679G/s encoded_file_size=51.8321M peak_memory_usage=710.717M
ParquetWrite/string_file_output/23/262143/1/0/0/manual_time         322 ms          275 ms            2 bytes_per_second=1.55169G/s encoded_file_size=277.944M peak_memory_usage=710.717M
ParquetWrite/string_file_output/23/524288/1/0/0/manual_time         436 ms          376 ms            2 bytes_per_second=1.14658G/s encoded_file_size=401.762M peak_memory_usage=733.125M
ParquetWrite/string_file_output/23/1048576/1/0/0/manual_time        529 ms          454 ms            1 bytes_per_second=967.006M/s encoded_file_size=527.884M peak_memory_usage=859.249M
ParquetWrite/string_file_output/23/2097152/1/0/0/manual_time        624 ms          534 ms            1 bytes_per_second=820.419M/s encoded_file_size=621.003M peak_memory_usage=937.209M
ParquetWrite/string_file_output/23/4194304/1/0/0/manual_time        633 ms          532 ms            1 bytes_per_second=809.252M/s encoded_file_size=656.618M peak_memory_usage=699.776M
```

**Comment by vuule** [comment]
@gpucibot merge

**Pull request merged**